### PR TITLE
Added option to replace disconnected clients with bots

### PR DESCRIPTION
**Pull request by mtarranger**
PR for [issue #3001](https://github.com/OpenRA/OpenRA/issues/3001): replacing disconnected clients with bots.

**Logic:**

Server drops player and send command "ActivateBot"
Game admin activates bot, notification in chat about this showed for everybody

**Detailed:**

(Option is enabled)
1. A player got dropped while a game is in progress.
2. Server doesn't send "player quit" packet (0xBF).
3. But it sends "Disconnected" instant message as usually - clients show "player disconnected" message.
4. Just after that server sends "ActivateBot" instant message.
- All clients (including admin) change their configuration so disconnected client is treated in the same way like it was a bot activated on the admin on start.
- It doesn't meter if dropped client had or had not sent packages for recent frames - clients will not check for them after this point (also look at explanation of third commit lower).
- All clients show "bot activated for <player name>" message
- Admin activates bot

Currently nothing will show that a player was real client and then was replaced by a bot except server chat message. Statistics will include combined player and bot progress after game end.

Please check spelling, especially for strings, variable names, code style, caveats.

I introduced:
"ActivateBot" command from server to clients
"Autopilot" string parameter in GlobalSettings which is a bot name if feature is enabled, null otherwise
"autopilot <bot name>" command from the admin to the lobby to enable option
"disable_autopilot" to disable

Please check implementation of the options gui. Do you have any cautions, will it work with specific maps (like scripted), foresights of bugs while player interact with the option in the lobby?

**Commits explanations (also look at the diffs and messages):**

Commit [1 of 4](https://github.com/OpenRA/OpenRA/commit/91491277ba9c3c21a0cb1a9b1d2bee18e2e5f00d):
- I didn't found any pattern for entries order for game options fields in lobby yaml, so I reordered them to follow left to right, line after line. That helped me in future changes.
- Some source and resource files had had execute bit on. It doesn't mean anything on windows, but that doesn't look pretty under *nix, so fixed some. On *nix you can check for that with `find -executable -type f` or `find -executable -type f -not -name '\*.sh' -not -path './.git/\*' -not -path './configure'` or auto fix with `find -executable -type f -name '\*.cs' | xargs chmod a-x` 

Commit [2 of 4](https://github.com/OpenRA/OpenRA/commit/626493676dbc47c272d56055d895e78c5b6f1b42):
- Added Session.Admin (LobbyInfo.Admin) as shortcut to `Clients.FirstOrDefault(c => c.IsAdmin)` to remove a small repetition.
- In OpenRA.Game/Server/Server.cs:DropClient() moved toDrop.Socket.Disconnect(false) at the beginning of the function ([#L525](https://github.com/mtarranger/OpenRA/blob/626493676dbc47c272d56055d895e78c5b6f1b42/OpenRA.Game/Server/Server.cs#L525)). As far as I can see it doesn't break anything, but this removes the feeling that if somehow processing will become multi-threaded, disconnecting client will have a possibility to change something after server prepared to disconnect him and actual disconnect.
- Mono doesn't respect HOME environment variable, and I didn't want to use Support directory. So I added OPENRASUPPORTDIR environment variable, which is path to already existing directory where logs, settings and other should be saved. Where should I write its description? Additionally there should be a command line option for this in future.

Commit [3 of 4 (network code)](https://github.com/OpenRA/OpenRA/commit/20e467ecb35a3f2d08964d4cb949350f4cf57785):
- In order to set up real client info like it is a bot client, client id should be removed from FrameData.clientQuitTimes ([#L35](https://github.com/mtarranger/OpenRA/blob/20e467ecb35a3f2d08964d4cb949350f4cf57785/OpenRA.Game/Network/FrameData.cs#L35)). This also excludes the client from IsReadyForNextFrame checks but doesn't affect FrameData.framePackets which holds any available order packets for disconnected client. The logic I chose implies that FrameData.clientQuitTimes is just optimization for checks who's playing and not for journaling purposes and can be freely modified.

Commit [4 of 4 (gui code)](https://github.com/OpenRA/OpenRA/commit/09bfb29ef33f15131fcb930b5133887c86029009):
Nothing special.

The projects code is very nice but it lacks comments (and self descriptive names doesn't always work).

IMHO network code and logic is a bit messy and requires some rewriting. The servers code becomes entangled.

Widgets logic code (at least in OpenRA.Mods.Common/Widgets/Logic/Lobby/) also a bit messy and a bit self repetitive.

**Conceptual questions:**
- Should autopilot be in map options? (Currently a map cannot set autopilot option)
- Can the removal of disconnected client id from FrameData.clientQuitTimes (described above in third commit) interfere with replay possibility? As it seems, replay will log "Disconnected", "ActivateBot" messages and will change FrameData.clientQuitTimes during replay as it happened during the game.
- Player name remain unchanged after the replacement (only chat message notes about replacement). I can change that if you want. In my understanding there are forces on the map that have names, and they are ruled by its command. I'd suggest that when a bot activated for a player bot type written in parenthesis after the player name, and when a bot added in the lobby it added with name "Bot".
- Also what should be BotControllerClientIndex if game doesn't have admins (dedicated, dropped)? I used `BotControllerClientIndex = admin == null ? 0 : admin.Index` ([#L90](https://github.com/mtarranger/OpenRA/blob/20e467ecb35a3f2d08964d4cb949350f4cf57785/OpenRA.Game/Network/UnitOrders.cs#L90)). Should it be -1 or other special?

**Random remarks** (that arose during the work):
- There are a couple of places where FirstOrDefault is used without further check if the result is null. Like in OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:LobbyLogic.LobbyLogic(), variable botController ([#L193](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs#L193), [#L576](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs#L576), changed from `LobbyInfo.Clients.FirstOrDefault(c => c.IsAdmin)` to `LobbyInfo.Admin` in my commits but it's the same) or in LobbyUtils.cs ([#L55](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs#L55))
- All "Send*" and "Dispatch*" in OpenRA.Game/Server/[Server.cs](https://github.com/OpenRA/OpenRA/blob/626493676dbc47c272d56055d895e78c5b6f1b42/OpenRA.Game/Server/Server.cs#L409) should be rethinked and rewritten.
- IsHost on clients and IsAdmin on server are ambiguous. At first thought IsHost should be true if client hosting a game, not dedicated server, and any one in the lobby can be selected as IsAdmin. Now they are used as the same thing.
- I would suggest to add in OpenRA.Game/Network/[FrameData.cs](https://github.com/mtarranger/OpenRA/blob/20e467ecb35a3f2d08964d4cb949350f4cf57785/OpenRA.Game/Network/FrameData.cs#L29),
  
  ```
  AddClient(clientId)
  {
  clientQuitTimes.Add(clientId, 0x7FFFFFFF)
  }
  ```
  
  as a way to tell FrameData that the clients packages is required for a next frame, and in OpenRA.Game/Network/[OrderManager.cs](https://github.com/mtarranger/OpenRA/blob/20e467ecb35a3f2d08964d4cb949350f4cf57785/OpenRA.Game/Network/OrderManager.cs#L85),
  
  ```
  WaitForClient(client)
  {
  frameData.AddClient(clientId)
  }
  ```
  
  for same reason, and remove `DispatchOrdersToClient(c, d.PlayerIndex, 0x7FFFFFFF, new byte[] { 0xBF });` from server (OpenRA.Game/Server/[Server.cs#L677](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Game/Server/Server.cs#L677)).
  
  But this should be a part of network code rewrite.
  
  WaitForClient can probably be called from one of GameStarted. Additionaly notifications (using interfaces) can be used for client disconnect/replacement.
- It's better to separate message strings from code (like in OpenRA.Mods.Common/Widgets/Logic/Lobby/[LobbyLogic.cs](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs#L414)), although I understand it should be done after internationalization will be implemented.
- Usage and meaning of Actor, IActor and others *Actor* in the project is a bit confusing, at least without comments (too many too different types of "actors").
- Order.PauseGame() (OpenRA.Game/Network/[Order.cs, #L150](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Game/Network/Order.cs#L150)) and PlayerCommands.InvokeCommand() (OpenRA.Mods.Common/Commands/[PlayerCommands.cs, #L43](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Commands/PlayerCommands.cs#L43)) contains same code `new Order("PauseGame", null, false) ...`. What's the right location? Also `new Order("Surrender", ...` in 2 places (OpenRA.Mods.Common/Commands/[PlayerCommands.cs, #L49](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Commands/PlayerCommands.cs#L49), OpenRA.Mods.Common/Widgets/Logic/Ingame/[IngameMenuLogic.cs, #L75](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs#L75))
- Flag numbers from network communication (like `0xBF`) can be put to constants with meaningful names (like FrameData.Markers.LastSeen or anything)
- OpenRA.Order.Deserialize() (OpenRA.Game/Network/[Order.cs, #L94](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Game/Network/Order.cs#L94)) and OpenRA.Server.ServerOrder.Deserialize() (OpenRA.Game/Server/[ServerOrder.cs, #L38](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Game/Server/ServerOrder.cs#L38)) contains same code. Maybe it's better to move read/write code to something like OrderFormat or change meaning of current names.
- Maybe strings in clients-server communication from OpenRA.Mods.Common ("TeamChat", "HandshakeRequest", "SyncInfo", ...) should be constants with special byte value (0x01 0x01, 0x01 0x02, ...)? Or at least string constants with current values and corresponding names to avoid typing mistakes?
- As it seems, meaning of Requires<T> (OpenRA.Game/Traits/TraitsInterfaces.cs) isn't documented anywhere, however it's usefull feature for mods.
- I think trait *.cs files in some directories can be placed more structured. Trait interfaces inside TraitsInterfaces.cs also (at least more sorted).
- Ending of OpenRA.Game/[Sync.cs](https://github.com/OpenRA/OpenRA/blob/b4fa704fbd61a8d66641f730adff5c9a43bdd498/OpenRA.Game/Sync.cs#L167) really needs comments. It looks like after the first successful fn execution inUnsyncedCode will never be false again.
- ./configure and make functionality is confusing. Typically configure creates Makefile configured to use tools available on platform and build with options set by configure. There is Makefile out of the box but compilation will fail because there is no Eluant.dll.config which is created by configure. `make` should fail with sensible message about configure wasn't executed or it should be changed to fulfill its task when called.
  I suggest either to remove configure at all and move its functionality to Makefile or rewrite configure so it will:
  - Detect platform libraries
  - Detect third-party libraries download method (is there nuget available or curl/wget should be used to download from `https://www.nuget.org/api/v2/package/$packageName/$packageVersion`. I did it with curl)
  - Write configs, scripts and Makefile
- I suggest OPENRAMODSDIR environment variable and command line option to point to custom mods directory location.
- Also at least short --help, -h and /? command line option would be nice.

### CODE REVIEW

**Comment by chrisforbes** [comment]
Thanks for the really detailed explanation of what this is trying to achieve. It's clear that some serious thought has gone into this.

I don't have time to review right now, but hopefully later this week.

**Comment by pchote** [comment]
I'm travelling for most of the next week, so also won't have time to give this a proper review. I gave it a quick skim, and two things stuck out:

* Double check that you aren't using an immediate order to send the activate bot command. Immediate orders are for non-world state (like chat etc) that should not be held up by the frame synchronization setup, and they will desync the game if they manipulate world state.

* Some maps explicitly disable bots, and this feature should respect that setting.

I also propose that our options for player disconnect should be "replace with bot" or "player surrenders" - leaving idle actors on the map is rather unpolished. I think we have a ticket for this, but it is a pain to search from my phone. If we do this, then the bot order probably wants to be replaced with a more general synced player disconnected order, and then a trait on the player actor can choose whether to activate the bot or destroy everything.

File: OpenRA.Game/Network/OrderManager.cs [hunk 1]
```diff
@@ -113,12 +121,17 @@ public void TickImmediate()
...
+					// Packet [int32 frame number][0xBF], sent by server when client dropped
 					if (packet.Length == 5 && packet[4] == 0xBF)
 						frameData.ClientQuit(clientId, frame);
+
 					else if (packet.Length >= 5 && packet[4] == 0x65)
 						CheckSync(packet);
+
 					else if (frame == 0)
 						immediatePackets.Add(Pair.New(clientId, packet));
+
```
Comment (Mailaender) [inline OpenRA.Game/Network/OrderManager.cs]: This is a StyleCop violation which makes our CI fail: [SA1510] An else, catch, or finally statement must not be preceded by a blank line.

**Comment by Mailaender** [comment]
Can you split the documentation and mode change commits into it's own pull request?

**Comment by penev92** [comment]
This hasn't seen any activity since it was open, so I'm closing it in at attempt to clean up the PR queue a bit. Please reopen it when you are able to continue your work :)

**Comment by MicroJOo** [comment]
#8369

**Comment by x-a-n-a-x** [comment]
What's the point of closing PRs just for the sake of clearing the PR queue? We all can live with 50 opened PRs but it seems you can't, penev. This is a bit ridiculous, iIMO. Just give developers time and the PR queue will feature less and let PRs slowly but surely.

**Comment by Mailaender** [comment]
Don't worry. @mtarranger can just click the reopen button himself. This is just helpful for us in an organisational way as we have to regularly check pull requests for silent updates and maybe add/remove those labels telling everyone what is left to do.

**Comment by WolfGaming** [comment]
@x-a-n-a-x Its so we dont spend the little reviewing time we have on PRs that arent active atm, having 1 active PR in the middle of 20 inactive ones makes the active one take longer.

**Comment by Mailaender** [comment]
> ./configure and make functionality is confusing.

We took care about that in https://github.com/OpenRA/OpenRA/pull/8082.

**Comment by Mailaender** [comment]
> Detect third-party libraries download method (is there nuget available or curl/wget should be used

I did that in https://github.com/OpenRA/OpenRA/pull/8229.

**Comment by Mailaender** [comment]
> Some source and resource files had had execute bit on.

Filed https://github.com/OpenRA/OpenRA/pull/8971 for that.

**Comment by x-a-n-a-x** [comment]
@mtarranger : is there a chance that (one day), you'll consider resuming your work on this PR ? 
It would be a great feature for OpenRA (and there is still a bounty on this issue.. ^^ )

**Comment by pchote** [comment]
This has been superseded by #10502, so removing from the milestone.

**Pull request closed without merging**